src/shared/telemetry/spans.ts

```diff
@@ -232,6 +260,19 @@ export class TelemetryTracer extends TelemetryBase {
         }
     }
 
+    /**
+     * Executes the given function within an anonymous 'root' span which does not emit
+     * any telemetry on its own.
+     *
+     * This can be used as a 'staging area' for adding attributes prior to creating spans.
```

Comment:
If runRoot() is called in multiple places , do all of their staged items get consumed by the first "real" span?

Reply:
No, the staging area only exists within the callback function, and so all spans _within the context_ will inherit the staged items. The purpose of `runRoot()` is to guarantee that we are in a context. This keeps the behavior of everything consistent: `record()` should behave the same regardless of your "parent" span.

Now I just need to change the behavior of `record()` to not back-propagate attributes.

Reply:
> When would a caller not call runRoot()? Why do the callers need to decide when to call run() vs runRoot()?

I'm basically only using `runRoot()` for the purpose of establishing `source`. It doesn't need to exist if we add definitions for these entry points and attach the parent span name to children spans. The majority of callers _shouldn't_ use `runRoot()`, it's just a means to defer adding more definitions.

> Also wondering if we can rename emit to runFoo (or vice-versa: run => emitFoo, or let emit take a function parameter which causes it to behave like run()) so that developers understand the basic equivalence of all these things.

Yeah `emit` was only added to avoid refactoring everything to use `run`. `emit` can be removed once things are converted.

Reply:
> Yeah `emit` was only added to avoid refactoring everything to use `run`. `emit` can be removed once things are converted.

Will help to mention practical stuff like that in the docstring on emit()